hail/src/main/scala/is/hail/expr/types/virtual/Type.scala

```diff
@@ -46,49 +42,31 @@ object Type {
         .toArray)
   }
 
-  def preGenStruct(required: Boolean, genFieldType: Gen[Type]): Gen[TStruct] = {
-    for (fields <- genFields(required, genFieldType)) yield {
-      val t = TStruct(fields)
-      if (required)
-        (+t).asInstanceOf[TStruct]
-      else
-        t
+  def preGenStruct(genFieldType: Gen[Type]): Gen[TStruct] = {
+    for (fields <- genFields(genFieldType)) yield {
+      TStruct(fields)
     }
   }
 
-  def preGenTuple(required: Boolean, genFieldType: Gen[Type]): Gen[TTuple] = {
-    for (fields <- genFields(required, genFieldType)) yield {
-      val t = TTuple(fields.map(_.typ): _*)
-      if (required)
-        (+t).asInstanceOf[TTuple]
-      else
-        t
+  def preGenTuple(genFieldType: Gen[Type]): Gen[TTuple] = {
+    for (fields <- genFields(genFieldType)) yield {
+      TTuple(fields.map(_.typ): _*)
     }
   }
 
   private val defaultRequiredGenRatio = 0.2
 
-  def genStruct: Gen[TStruct] = Gen.coin(defaultRequiredGenRatio).flatMap(preGenStruct(_, genArb))
-
-  val genOptionalStruct: Gen[Type] = preGenStruct(required = false, genArb)
+  def genStruct: Gen[TStruct] = preGenStruct(genArb)
 
-  val genRequiredStruct: Gen[Type] = preGenStruct(required = true, genArb)
-
-  val genInsertableStruct: Gen[TStruct] = Gen.coin(defaultRequiredGenRatio).flatMap(required =>
-    if (required)
-      preGenStruct(required = true, genArb)
-    else
-      preGenStruct(required = false, genOptional))
-
-  def genSized(size: Int, required: Boolean, genTStruct: Gen[TStruct]): Gen[Type] =
+  def genSized(size: Int, genTStruct: Gen[TStruct]): Gen[Type] =
     if (size < 1)
-      Gen.const(TStruct.empty(required))
+      Gen.const(TStruct.empty)
     else if (size < 2)
-      genScalar(required)
+      genScalar()
     else {
       Gen.frequency(
-        (4, genScalar(required)),
-        (1, genComplexType(required)),
+        (4, genScalar()),
```

Comment:
these can be called without parens

Reply:
I kinda hate no-parens defs. They make it so easy to inline code you think is evaluated once.

Reply:
sure